### PR TITLE
Discrete calculus model of Ambrosio Tortorelli functional

### PR DESCRIPTION
**Pull request by JacquesOlivierLachaud**
# PR Description

This PR brings the 2D discrete calculus model of Ambrosio-Tortorelli functional into DGtal DEC package. It is useful to compute piecewise-smooth approximation of scalar or vector fields defined on a 2D grid or a 2D digital surface in arbitrary dimension.

# Checklist

- [x] Doxygen documentation of the code completed (classes, methods, types, members...)
- [x] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [x] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

### CODE REVIEW

File: src/DGtal/helpers/Shortcuts.h [hunk 1]
```diff
@@ -1373,6 +1373,68 @@ namespace DGtal
...
+      /// @return a range of cells as a vector.
+      template <typename TDigitalSurfaceContainer>
+      static CellRange
+      getCellRange
+      ( Cell2Index& c2i,
+        CountedPtr< ::DGtal::DigitalSurface<TDigitalSurfaceContainer> > surface,
+        const Dimension k )
+      {
+        CellRange result;
+        result.reserve( 2 * surface->size() + 100 );
```
Comment (dcoeurjo) [inline src/DGtal/helpers/Shortcuts.h]: where does the +100 come from?

Comment (dcoeurjo) [reply src/DGtal/helpers/Shortcuts.h]: well.. the overall formula BTW

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 2]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+      ///   - at-alpha        [  0.1   ]: parameter alpha in AT (data fit)
+      ///   - at-lambda       [  0.025 ]: parameter lambda in AT (1/length of discontinuities)
+      ///   - at-epsilon      [  0.5   ]: (last value of) parameter epsilon in AT (width of discontinuities)
+      ///   - at-epsilon-start[  2.0   ]: first value for parameter epsilon in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-epsilon-ratio[  2.0   ]: ratio between two consecutive epsilon value in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-max-iter     [ 10     ]: maximum number of alternate minization in AT optimization
+      ///   - at-diff-v-max   [  0.0001]: stopping criterion that measures the loo-norm of the evolution of \a v between two iterations
+      ///   - at-v-policy     ["Maximum"]: the policy when outputing feature vector v onto cells: "Average"|"Minimum"|"Maximum"
+      ///
+      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN ..`
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]:
```suggestion
      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN=true ...`
```

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 3]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+          ( "at-v-policy",   "Maximum" );
+#else // defined(WITH_EIGEN)
+        return Parameters( "at-enabled", 0 );
+#endif// defined(WITH_EIGEN)
+      }
+
+#if defined(WITH_EIGEN)
+
+      /// Given any digital \a surface, a surfel range \a surfels, and an input vector field \a input,
+      /// returns a piece-smooth approximation of \a input using Ambrosio-Tortorelli functional.
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]: ref to the module page ?

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 4]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+      ///   - at-epsilon      [  0.5   ]: (last value of) parameter epsilon in AT (width of discontinuities)
+      ///   - at-epsilon-start[  2.0   ]: first value for parameter epsilon in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-epsilon-ratio[  2.0   ]: ratio between two consecutive epsilon value in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-max-iter     [ 10     ]: maximum number of alternate minization in AT optimization
+      ///   - at-diff-v-max   [  0.0001]: stopping criterion that measures the loo-norm of the evolution of \a v between two iterations
+      /// @param[in] input the input vector field (a vector of vector values)
+      ///
+      /// @return the piecewise-smooth approximation of \a input.
+      ///
+      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN ..`
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]:
```suggestion
      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN=true ...`
```

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 5]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+      ///   - at-epsilon-start[  2.0   ]: first value for parameter epsilon in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-epsilon-ratio[  2.0   ]: ratio between two consecutive epsilon value in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-max-iter     [ 10     ]: maximum number of alternate minization in AT optimization
+      ///   - at-diff-v-max   [  0.0001]: stopping criterion that measures the loo-norm of the evolution of \a v between two iterations
+      ///   - at-v-policy     ["Maximum"]: the policy when outputing feature vector v onto cells: "Average"|"Minimum"|"Maximum"
+      /// @param[in] input the input vector field (a vector of vector values)
+      ///
+      /// @return the piecewise-smooth approximation of \a input.
+      ///
+      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN ..`
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]:
```suggestion
      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN=true ...`
```

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 6]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+        at_solver.initInputVectorFieldU2( input, surfels.cbegin(), surfels.cend() );
+        at_solver.setUp( alpha_at, lambda_at );
+        at_solver.solveGammaConvergence( epsilon1, epsilon2, epsilonr, false, diff_v_max, max_iter );
+        auto output = input;
+        at_solver.getOutputVectorFieldU2( output, surfels.cbegin(), surfels.cend() );
+        return output;
+      }
 
+      /// Given any digital \a surface, a surfel range \a surfels, and an input vector field \a input,
+      /// returns a piece-smooth approximation of \a input using Ambrosio-Tortorelli functional.
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]: ref module page

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 7]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+      ///   - at-epsilon      [  0.5   ]: (last value of) parameter epsilon in AT (width of discontinuities)
+      ///   - at-epsilon-start[  2.0   ]: first value for parameter epsilon in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-epsilon-ratio[  2.0   ]: ratio between two consecutive epsilon value in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-max-iter     [ 10     ]: maximum number of alternate minization in AT optimization
+      ///   - at-diff-v-max   [  0.0001]: stopping criterion that measures the loo-norm of the evolution of \a v between two iterations
+      /// @param[in] input the input scalar field (a vector of scalar values)
+      ///
+      /// @return the piecewise-smooth approximation of \a input.
+      ///
+      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN ..`
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]:
```suggestion
      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN=true ...`
```

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 8]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+      ///   - at-epsilon-start[  2.0   ]: first value for parameter epsilon in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-epsilon-ratio[  2.0   ]: ratio between two consecutive epsilon value in Gamma-convergence optimization (sequence of AT optimization with decreasing epsilon)
+      ///   - at-max-iter     [ 10     ]: maximum number of alternate minization in AT optimization
+      ///   - at-diff-v-max   [  0.0001]: stopping criterion that measures the loo-norm of the evolution of \a v between two iterations
+      ///   - at-v-policy     ["Maximum"]: the policy when outputing feature vector v onto cells: "Average"|"Minimum"|"Maximum"
+      /// @param[in] input the input scalar field (a vector of scalar values)
+      ///
+      /// @return the piecewise-smooth approximation of \a input.
+      ///
+      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN ..`
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]:
```suggestion
      /// @note Requires Eigen linear algebra backend. `Use cmake -DWITH_EIGEN=true ...`
```

File: src/DGtal/helpers/ShortcutsGeometry.h [hunk 9]
```diff
@@ -965,11 +975,293 @@ namespace DGtal
...
+        at_solver.solveGammaConvergence( epsilon1, epsilon2, epsilonr, false, diff_v_max, max_iter );
+        auto output = input;
+        at_solver.getOutputScalarFieldU2( output, surfels.cbegin(), surfels.cend() );
+        return output;
+      }
+
+      /// Given any digital \a surface, a surfel range \a surfels, and
+      /// an input scalar field \a input, returns a piece-smooth
+      /// approximation of \a input using Ambrosio-Tortorelli
+      /// functional.  Given a range of pointels, linels or 2-cells
```
Comment (dcoeurjo) [inline src/DGtal/helpers/ShortcutsGeometry.h]: ref module page

File: src/DGtal/dec/ATSolver2D.h [hunk 10]
```diff
@@ -0,0 +1,1028 @@
...
+  /**
+  * Description of template class 'ATSolver2D' <p> \brief Aim:
+  * This class solves Ambrosio-Tortorelli functional on a
+  * two-dimensional digital space (a 2D grid or 2D digital surface)
+  * for a piecewise smooth scalar/vector function \a u represented as
+  * one/several 2-form(s) and a discontinuity function \a v
+  * represented as a 0-form. The 2-form(s) \a u is a regularized
+  * approximation of an input vector data \a g, while \a v represents
+  * the set of discontinuities of \a u.  The norm chosen for \a u is
+  * the l2-norm.
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]:
```suggestion
  * the \f$ l_2\f$-norm.
```

File: src/DGtal/dec/ATSolver2D.h [hunk 11]
```diff
@@ -0,0 +1,1028 @@
...
+  * typedef DiscreteExteriorCalculusFactory<EigenLinearAlgebraBackend> CalculusFactory;
+  * const auto calculus = CalculusFactory::createFromNSCells<2>( surfels.begin(), surfels.end() );
+  * ATSolver2D< KSpace > at_solver(calculus, 1);
+  * at_solver.initInputVectorFieldU2( normals, surfels.cbegin(), surfels.cend() );
+  * at_solver.setUp( alpha_at, lambda_at );
+  * at_solver.solveGammaConvergence( 2.0, 0.5, 2.0 );
+  * at_solver.getOutputVectorFieldU2( normals, surfels.cbegin(), surfels.cend() );
+  * \endcode
+  *
+  * @see exampleSurfaceATNormals.cpp
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: ref not working.. strange..

File: src/DGtal/dec/ATSolver2D.h [hunk 12]
```diff
@@ -0,0 +1,1028 @@
...
+    PrimalDerivative0     primal_D0;
+    /// the derivative operator for primal 1-forms
+    PrimalDerivative1     primal_D1;
+    /// The primal vertex to edge average operator
+    PrimalDerivative0     M01; 
+    /// The primal edge to face average operator
+    PrimalDerivative1     M12; 
+    /// The antiderivative of primal 2-forms
+    PrimalAntiderivative2 primal_AD2;
+    /// The alpha-weighted identity operator for primal 2-forms (stored for performance)
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: maybe not exposed in `public:`?

File: src/DGtal/dec/ATSolver2D.h [hunk 13]
```diff
@@ -0,0 +1,1028 @@
...
+    // typedef EigenLinearAlgebraBackend::SolverSparseLU LinearAlgebraSolver;
+    // typedef EigenLinearAlgebraBackend::SolverSimplicialLLT LinearAlgebraSolver;
+    typedef EigenLinearAlgebraBackend::SolverSimplicialLDLT LinearAlgebraSolver;
+    typedef DiscreteExteriorCalculusSolver<Calculus, LinearAlgebraSolver, 2, PRIMAL, 2, PRIMAL> SolverU2;
+    typedef DiscreteExteriorCalculusSolver<Calculus, LinearAlgebraSolver, 0, PRIMAL, 0, PRIMAL> SolverV0;
+
+    /// A smart (or not) pointer to a calculus object.
+    CountedConstPtrOrConstPtr< Calculus > ptrCalculus;
+    /// the derivative operator for primal 0-forms
+    PrimalDerivative0     primal_D0;
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: shouldn't all these be `private:` members ?

File: src/DGtal/dec/ATSolver2D.h [hunk 14]
```diff
@@ -0,0 +1,1028 @@
...
+    /// Solves the alternate minimization of AT for a given \a eps. Solves
+    /// for u then for v till convergence.
+    ///
+    /// @param eps the epsilon parameter at which AT is solved.
+    ///
+    /// @param n_oo_max the alternate minimization will stop when the
+    /// loo-norm of \f$ v^{k+1} - v^k \f$ is below this bound.
+    ///
+    /// @param iter_max the alternate minimization will stop when the
+    /// number of minimization steps exceed \a iter_max.
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]:
```suggestion
    /// number of minimization steps exceeds \a iter_max.
```

File: src/DGtal/dec/ATSolver2D.h [hunk 15]
```diff
@@ -0,0 +1,1028 @@
...
+    /// @param eps1 the first epsilon parameter at which AT is solved.
+    /// @param eps2 the last epsilon parameter at which AT is solved.
+    /// @param epsr the ratio (>1) used to decrease progressively epsilon.
+    /// @param compute_smallest_epsilon_map when 'true' determines for each surfel the smallest epsilon for which it is a discontinuity.
+    ///
+    /// @param n_oo_max the alternate minimization will stop when the
+    /// loo-norm of \f$ v^{k+1} - v^k \f$ is below this bound.
+    ///
+    /// @param iter_max the alternate minimization will stop when the
+    /// number of minimization steps exceed \a iter_max.
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]:
```suggestion
    /// number of minimization steps exceeds \a iter_max.
```

and elsewhere

File: src/DGtal/dec/ATSolver2D.h [hunk 16]
```diff
@@ -0,0 +1,1028 @@
...
+              case CellOutputPolicy::Maximum:
+                output[ i ] = std::max( std::max( v0.myContainer( idx00 ), v0.myContainer( idx01 ) ),
+                                        std::max( v0.myContainer( idx10 ), v0.myContainer( idx11 ) ) );
+                break;
+              }
+            }
+        }
+    }
+    
+    /// Computes the map that tells for each surfel the smallest
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: the sentence is a bit weird...

File: src/DGtal/dec/ATSolver2D.h [hunk 17]
```diff
@@ -0,0 +1,1028 @@
...
+     */
+    bool isValid() const
+    {
+      return true;
+    }
+
+    /// @}
+    
+    // ------------------------- Protected Datas ------------------------------
+  protected:
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: cleanup

File: src/DGtal/dec/ATSolver2D.h [hunk 18]
```diff
@@ -0,0 +1,1028 @@
...
+
+    // ------------------------- Private Datas --------------------------------
+  private:
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: cleanup

File: src/DGtal/dec/ATSolver2D.h [hunk 19]
```diff
@@ -0,0 +1,1028 @@
...
+    /// @param[out] output the vector of output values (a scalar field), which should be of size `length(itB,itE)`
+    ///
+    /// @param itB the start of the range of surfels.
+    /// @param itE past the end of the range of surfels.
+    template <typename ScalarFieldOutput, typename SurfelRangeConstIterator>
+    void
+    getOutputScalarFieldU2( ScalarFieldOutput& output,
+                            SurfelRangeConstIterator itB, SurfelRangeConstIterator itE )
+    {
+      const Dimension N = u2.size();
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: this will raise "unused variable" warnings in non Debug mode..

File: src/DGtal/dec/ATSolver2D.h [hunk 20]
```diff
@@ -0,0 +1,1028 @@
...
+      }
+      setEpsilon( eps );
+      for ( unsigned int i = 0; i < iter_max; ++i )
+        {
+	  if ( verbose >= 1 )
+	    trace.info() << "---------- Iteration "
+			 << i << "/" << iter_max << " ---------------" << std::endl;
+	  solveOneAlternateStep( );
+	  //bool ok = ok && 
+	  auto norms_v = checkV0();
```
Comment (dcoeurjo) [inline src/DGtal/dec/ATSolver2D.h]: unused ?

File: src/DGtal/dec/doc/packageDEC.dox [hunk 21]
```diff
@@ -56,6 +56,7 @@ Basic operators, such as Hodge duality operator or exterior derivative, can be c
 - \ref moduleDECHelmoltz
 - \ref moduleDECPropagation
 - \ref moduleHeatLaplaceOperator
+- \ref moduleAT
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/doc/packageDEC.dox]: I think I have added the authors in the module page (the four of us: PG, MF, you and myself). What do you mean ?

File: src/DGtal/dec/doc/moduleAT.dox [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+
+@writers Jacques-Olivier Lachaud, Marion Foare, David Coeurjolly, Pierre Gueth
+
+@since 1.1
+
+  Part of package \ref packageDEC.
+  [TOC]
+
+- related classes: \ref ATSolver2D
+- related examples: \ref exampleSurfaceATNormals.cpp
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/doc/moduleAT.dox]: Weird. It works for me when I build the doc locally. Is there a place where I can see the resulting doc by CI ?

File: src/DGtal/dec/doc/moduleAT.dox [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+<td> \image html at-fandisk-128-vf-normals-features.png "Piecewise smooth normal vectors (sticks) extracted by AT model" </td>
+</tr>
+</table>
+
+To detail a little bit how to do that, you have first to build a
+digital surface. For instance, you can use Shortcuts to do so quickly
+(see \ref moduleShortcuts):
+
+\code
+#include "ConfigExamples.h"
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/doc/moduleAT.dox]: Done

File: src/DGtal/dec/doc/moduleAT.dox [hunk 24]
```diff
@@ -0,0 +1,258 @@
...
+
+After that, the AT solver can be build on this calculus. Then it is
+initialized with input data you wish to approximate (here the normal
+vector field \a ii_normals). Finally it suffices to call the method
+ATSolver2D::solveGammaConvergence to use the coarse-to-fine \f$ \Gamma
+\f$-convergence optimization method.
+
+\code
+const double alpha_at  = 0.1;   // data fit
+const double lambda_at = 0.025; // 1/length of discontinuities
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/doc/moduleAT.dox]: done

[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/doc/moduleAT.dox]: done

File: src/DGtal/helpers/Shortcuts.h (hunk 1 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/Shortcuts.h]: Approximate number of surfels for a 2d surface. I have put a comment in the code

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 2 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: done

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 3 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: done

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 4 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: done

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 6 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: yes

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 7 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: done

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 9 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: done

File: src/DGtal/dec/ATSolver2D.h (hunk 10 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: done

File: src/DGtal/dec/ATSolver2D.h (hunk 12 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: ok

File: src/DGtal/dec/ATSolver2D.h (hunk 13 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: ok

File: src/DGtal/dec/ATSolver2D.h (hunk 14 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: done

File: src/DGtal/dec/ATSolver2D.h (hunk 15 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: done

File: src/DGtal/dec/ATSolver2D.h (hunk 16 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: I tried to fix it

File: src/DGtal/dec/ATSolver2D.h (hunk 18 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: done

File: src/DGtal/dec/ATSolver2D.h (hunk 17 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: done

File: src/DGtal/dec/ATSolver2D.h (hunk 19 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: You are right. Fixed

File: src/DGtal/dec/ATSolver2D.h (hunk 20 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/ATSolver2D.h]: removed

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 8 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: done

File: src/DGtal/helpers/ShortcutsGeometry.h (hunk 5 above)
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/ShortcutsGeometry.h]: done

File: src/DGtal/dec/doc/moduleAT.dox (hunk 22 above)
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/doc/moduleAT.dox]: Seems fixed locally

**Comment by dcoeurjo** [comment]
For the authors: in some package pages, module page authors are copied in between parenthesis. This is not the case for DEC (only Pierre before this PR)

File: src/DGtal/dec/doc/moduleAT.dox (hunk 22 above)
[the remark this answers is not in the document]
Comment (dcoeurjo) [reply src/DGtal/dec/doc/moduleAT.dox]: I also tested it locally. Weird

**Comment by JacquesOlivierLachaud** [comment]
I can add authors in parenthesis in the package if you wish.
By the way, for me an author is someone who contributed somehow to the module, not necessarily one who has produced code (you can see how I acknowledge people in several modules). In this case, the writers of the code may not be the same (here I will also put the four of us, but anyway you got the idea).

**Comment by dcoeurjo** [comment]
Agree (as we did for other package)

**Comment by JacquesOlivierLachaud** [comment]
Everything should be fine now

File: src/DGtal/helpers/Shortcuts.h [hunk 25]
```diff
@@ -1490,38 +1552,38 @@ namespace DGtal
...
+        const Parameters&   params = parametersDigitalSurface() )
+      {
+        typedef ::DGtal::DigitalSurface<TDigitalSurfaceContainer> AnyDigitalSurface;
+        SurfelRange result;
+        std::string traversal = params[ "surfaceTraversal" ].as<std::string>();
+        if ( traversal == "DepthFirst" )
+          {
+            typedef DepthFirstVisitor< AnyDigitalSurface > Visitor;
+            typedef GraphVisitorRange< Visitor > VisitorRange;
+            VisitorRange range( new Visitor( *surface, start_surfel ) );
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/Shortcuts.h]: It is acquired and stored in a CountedPtr, so no worries as would say the Aussies.

File: src/DGtal/helpers/Shortcuts.h [hunk 26]
```diff
@@ -1490,38 +1552,38 @@ namespace DGtal
...
+            std::for_each( range.begin(), range.end(),
+                           [&result] ( Surfel s ) { result.push_back( s ); } );
+          }
+        else if ( traversal == "BreadthFirst" )
+          {
+            typedef BreadthFirstVisitor< AnyDigitalSurface > Visitor;
+            typedef GraphVisitorRange< Visitor > VisitorRange;
+            VisitorRange range( new Visitor( *surface, start_surfel ) );
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/helpers/Shortcuts.h]: Same as above.

File: src/DGtal/dec/DECHelpers.h [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+              const Cell surfel = kspace.uCell((kspace.uKCoords(point)+delta));
+              ASSERT( kspace.uDim(surfel) == 2 );
+              if (calculus.containsCell(surfel))
+                indexes_surfel.push_back(calculus.getCellIndex(surfel));
+            }
+          ASSERT( indexes_surfel.size() > 2 );
+          
+          const double weight = 1/static_cast<Scalar>(indexes_surfel.size());
+          for (const Index index_surfel : indexes_surfel)
+            triplets.push_back(Triplet(index_point, index_surfel, weight));
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/DECHelpers.h]: agree

File: src/DGtal/dec/DECHelpers.h [hunk 28]
```diff
@@ -0,0 +1,172 @@
...
+    diagonal(const DGtal::KForm<Calculus, dim, duality>& kform)
+    {
+      typedef DGtal::LinearOperator<Calculus,dim, duality, dim, duality> Operator;
+      typedef typename Calculus::LinearAlgebraBackend::Triplet Triplet;
+      typedef typename Calculus::Index Index;
+      typedef std::vector<Triplet> Triplets;
+
+      Triplets triplets;
+      for (Index index=0; index<kform.length(); index++)
+        triplets.push_back(Triplet(index, index, kform.myContainer(index)));
```
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/DECHelpers.h]: agree

File: src/DGtal/dec/DECHelpers.h (hunk 27 above)
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/DECHelpers.h]: done

File: src/DGtal/dec/DECHelpers.h (hunk 28 above)
[the remark this answers is not in the document]
Comment (JacquesOlivierLachaud) [reply src/DGtal/dec/DECHelpers.h]: done

**Comment by dcoeurjo** [comment]
thanks a lot @JacquesOlivierLachaud & @copyme 
merging

**Pull request merged**